File: patterns/integration/aggregator.py
"""
Aggregator Pattern - Python

Combines multiple messages/requests into a single response,
reducing network overhead and improving performance.
"""

import threading
import time


class MessageAggregator:
    def __init__(self, timeout=1.0, max_size=10):
        self.timeout = timeout
        self.max_size = max_size
        self.buffer = []
        self.timer = None
        self.lock = threading.Lock()
    
    def add(self, message, callback):
        with self.lock:
            self.buffer.append({'message': message, 'callback': callback})
            
            if len(self.buffer) >= self.max_size:
                self.flush()
            elif not self.timer:
                self.timer = threading.Timer(self.timeout, self.flush)
                self.timer.start()
    
    def flush(self):
        with self.lock:
            if self.timer:
                self.timer.cancel()
                self.timer = None
            
            if self.buffer:
                batch = self.buffer[:]
                self.buffer.clear()
                
                aggregated = self.aggregate([item['message'] for item in batch])
                
                for item in batch:
                    item['callback'](aggregated)
    
    def aggregate(self, messages):
        return {
            'count': len(messages),
            'messages': messages,
            'timestamp': time.time()
        }


# Example usage
if __name__ == "__main__":
    print("=== Aggregator Pattern ===\n")
    
    aggregator = MessageAggregator(0.5, 5)
    
    for i in range(1, 8):
        aggregator.add(
            {'id': i, 'data': f'Message {i}'},
            lambda aggregated: print(f"Received aggregated batch: {aggregated['count']} messages")
        )
    
    time.sleep(1)
    print("Done")

File: patterns/integration/aggregator.js
/**
 * Aggregator Pattern - JavaScript
 * 
 * Combines multiple messages/requests into a single response,
 * reducing network overhead and improving performance.
 */

class MessageAggregator {
    constructor(timeout = 1000, maxSize = 10) {
        this.timeout = timeout;
        this.maxSize = maxSize;
        this.buffer = [];
        this.timer = null;
    }
    
    add(message, callback) {
        this.buffer.push({ message, callback });
        
        if (this.buffer.length >= this.maxSize) {
            this.flush();
        } else if (!this.timer) {
            this.timer = setTimeout(() => this.flush(), this.timeout);
        }
    }
    
    flush() {
        if (this.timer) {
            clearTimeout(this.timer);
            this.timer = null;
        }
        
        if (this.buffer.length > 0) {
            const batch = this.buffer.splice(0);
            const aggregated = this.aggregate(batch.map(item => item.message));
            
            batch.forEach(item => {
                item.callback(aggregated);
            });
        }
    }
    
    aggregate(messages) {
        return {
            count: messages.length,
            messages: messages,
            timestamp: new Date()
        };
    }
}

// Example usage
console.log('=== Aggregator Pattern ===\n');

const aggregator = new MessageAggregator(500, 5);

for (let i = 1; i <= 7; i++) {
    aggregator.add(
        { id: i, data: `Message ${i}` },
        (aggregated) => {
            console.log(`Received aggregated batch: ${aggregated.count} messages`);
        }
    );
}

// Wait for flush
setTimeout(() => {
    console.log('Done');
}, 1000);

module.exports = { MessageAggregator };

